File: ios/Classes/MxZendeskMessaging.swift
//
//  MxZendeskMessaging.swift
//  mx_zendesk_messaging
//
//  Created by Magical Water on 2023/5/5.
//

import Flutter
import UIKit
import ZendeskSDKMessaging
import ZendeskSDK

public class MxZendeskMessaging: NSObject {
    let TAG = "[MxZendeskMessaging]"
    
    private var channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    func initialize(channelKey: String, callId: Int?, onComplete: @escaping (Bool) -> Void) {
        printLog("\(self.TAG) - Channel Key - \(channelKey)\n")
        Zendesk.initialize(withChannelKey: channelKey, messagingFactory: DefaultMessagingFactory()) { result -> Void in
            
            var argsMap: [String: Any] = ["callId": callId!]
            if case let .failure(error) = result {
                onComplete(false)
                self.printLog("\(self.TAG) - initialize failure - \(error.localizedDescription)\n")
                argsMap["error"] = error.localizedDescription
            } else {
                // 初始化成功後加入事件觀察
                self.setEventObservable()
                onComplete(true)
                self.printLog("\(self.TAG) - initialize success")
            }
            
            self.channel.invokeMethod(ZendeskChannelMethod.initialize.rawValue, arguments: argsMap)
        }
    }
    
    func show(rootViewController: UIViewController?, callId: Int?) {
        
        let messagingViewController = Zendesk.instance?.messaging?.messagingViewController()
        
        var argsMap: [String: Any] = ["callId": callId!]
        
        if (rootViewController == nil) {
            argsMap["error"] = "rootViewController is nil"
        } else if (messagingViewController == nil) {
            argsMap["error"] = "messagingViewController is nil"
        } else {
            rootViewController!.present(messagingViewController!, animated: true, completion: nil)
        }
        
        printLog("\(self.TAG) - show")
        channel.invokeMethod(ZendeskChannelMethod.show.rawValue, arguments: argsMap)
    }
    
    func loginUser(jwt: String, callId: Int?) {
        Zendesk.instance?.loginUser(with: jwt) { result in
            var argsMap: [String: Any] = ["callId": callId!]
            switch result {
            case .success(let user):
                argsMap["id"] = user.id
                argsMap["externalId"] = user.externalId
                break
            case .failure(let error):
                argsMap["error"] = error.localizedDescription
                self.printLog("\(self.TAG) - login failure - \(error.localizedDescription)\n")
                break
            }
            
            self.channel.invokeMethod(ZendeskChannelMethod.loginUser.rawValue, arguments: argsMap)
        }
    }
    
    func logoutUser(callId: Int?) {
        Zendesk.instance?.logoutUser { result in
            var argsMap: [String: Any] = ["callId": callId!]
            
            switch result {
            case .success:
                break
            case .failure(let error):
                argsMap["error"] = error.localizedDescription
                self.printLog("\(self.TAG) - logout failure - \(error.localizedDescription)\n")
                break
            }
            
            self.channel.invokeMethod(ZendeskChannelMethod.logoutUser.rawValue, arguments: argsMap)
        }
    }
    
    func getUnreadMessageCount() -> Int {
        return Zendesk.instance?.messaging?.getUnreadMessageCount() ?? 0
    }
    
    // 設置事件觀察
    func setEventObservable() {
        removeEventObservable()
        printLog("\(self.TAG) - 加入事件觀察 \(Zendesk.instance != nil)")
        Zendesk.instance?.addEventObserver(self) { event in
            switch event {
            case .unreadMessageCountChanged(let unreadCount):
                self.channel.invokeMethod(ZendeskChannelMethod.eventUnreadCount.rawValue, arguments: ["content": unreadCount])
                break
            case .authenticationFailed(let error as NSError):
                print("Authentication error received: \(error)")
                print("Domain: \(error.domain)")
                print("Error code: \(error.code)")
                print("Localized Description: \(error.localizedDescription)")
                self.channel.invokeMethod(ZendeskChannelMethod.eventAuthFail.rawValue, arguments: ["code": error.code, "domain": error.domain, "error": error.localizedDescription])
                break
            @unknown default:
                self.channel.invokeMethod(ZendeskChannelMethod.eventUnknown.rawValue, arguments: ["content": "\(event)"])
                break
            }
        }
    }
    
    func removeEventObservable() {
        Zendesk.instance?.removeEventObserver(self)
    }
    
    func printLog(_ text: String) {
        channel.invokeMethod(ZendeskChannelMethod.printLog.rawValue, arguments: ["content": text])
    }
}

File: android/src/main/kotlin/org/water/mx_zendesk_messaging/MxZendeskMessaging.kt
package org.water.mx_zendesk_messaging

import android.content.Context
import android.content.Intent
import io.flutter.plugin.common.MethodChannel
import zendesk.android.Zendesk
import zendesk.android.ZendeskResult
import zendesk.android.ZendeskUser
import zendesk.android.events.ZendeskEvent
import zendesk.android.events.ZendeskEventListener
import zendesk.messaging.android.DefaultMessagingFactory

class MxZendeskMessaging(
    private val channel: MethodChannel,
) {
    companion object {
        const val tag = "[MxZendeskMessaging]"
    }

    fun initialize(
        context: Context,
        channelKey: String,
        callId: Int,
        onComplete: (Boolean) -> Unit
    ) {
        printLog("$tag - Channel Key - $channelKey")
        Zendesk.initialize(context, channelKey, successCallback = { value ->
            // 初始化成功後加入事件觀察
            setEventObservable()
            val args = mapOf<String, Any?>("callId" to callId)
            onComplete(true)
            printLog("$tag - initialize success - $value")
            channel.invokeMethod(ZendeskChannelMethod.Initialize.rawValue, args)
        }, failureCallback = { error ->
            val args = mapOf<String, Any?>("callId" to callId, "error" to error.message)
            onComplete(false)
            printLog("$tag - initialize failure - $error")
            channel.invokeMethod(ZendeskChannelMethod.Initialize.rawValue, args)
        }, messagingFactory = DefaultMessagingFactory()
        )
    }

    fun show(context: Context, callId: Int?) {
        val args = mutableMapOf<String, Any?>("callId" to callId)
        printLog("$tag - show")
        Zendesk.instance.messaging.showMessaging(context, Intent.FLAG_ACTIVITY_NEW_TASK)
        channel.invokeMethod(ZendeskChannelMethod.Show.rawValue, args)
    }

    fun getUnreadMessageCount(): Int {
        return Zendesk.instance.messaging.getUnreadMessageCount()
    }

    fun loginUser(jwt: String, callId: Int) {
        Zendesk.instance.loginUser(jwt, { value: ZendeskUser? ->
            val args = mapOf<String, Any?>(
                "callId" to callId,
                "id" to value?.id,
                "externalId" to value?.externalId
            )
            channel.invokeMethod(ZendeskChannelMethod.LoginUser.rawValue, args)
        }, { error: Throwable? ->
            val args = mapOf<String, Any?>(
                "callId" to callId,
                "error" to error?.message
            )
            printLog("$tag - Login failure : ${error?.message}")

            channel.invokeMethod(ZendeskChannelMethod.LoginUser.rawValue, args)
        })
    }

    suspend fun logoutUser(callId: Int) {
        val args: MutableMap<String, Any?> = mutableMapOf("callId" to callId)
        val result = Zendesk.instance.logoutUser()
        if (result is ZendeskResult.Failure) {
            args["error"] = result.error.message
        }
        channel.invokeMethod(ZendeskChannelMethod.LogoutUser.rawValue, args)
    }

    // 事件監聽
    private val zendeskEventListener: ZendeskEventListener = ZendeskEventListener { zendeskEvent ->
        when (zendeskEvent) {
            is ZendeskEvent.UnreadMessageCountChanged -> {
                channel.invokeMethod(
                    ZendeskChannelMethod.EventUnreadCount.rawValue,
                    mapOf("content" to zendeskEvent.currentUnreadCount)
                )
            }
            is ZendeskEvent.AuthenticationFailed -> {
                channel.invokeMethod(
                    ZendeskChannelMethod.EventAuthFail.rawValue,
                    mapOf("error" to zendeskEvent.error.message)
                )
            }
            else -> {
                channel.invokeMethod(
                    ZendeskChannelMethod.EventUnknown.rawValue,
                    mapOf("content" to zendeskEvent.toString())
                )
            }
        }
    }

    // 設置事件觀察
    fun setEventObservable() {
        removeEventObservable()
        printLog("$tag - 加入事件觀察")
        Zendesk.instance.addEventListener(zendeskEventListener)
    }

    private fun removeEventObservable() {
        Zendesk.instance.removeEventListener(zendeskEventListener)
    }

    private fun printLog(text: Any?) {
        channel.invokeMethod(
            ZendeskChannelMethod.PrintLog.rawValue, mapOf("content" to text?.toString())
        )
    }
}


